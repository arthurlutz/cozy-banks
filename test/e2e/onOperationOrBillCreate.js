process.env.NODE_ENV = 'development'

const fs = require('fs')
const path = require('path')
const {
  runService,
  toMatchSnapshot,
  ach,
  makeToken,
  prompt,
  couch
} = require('./utils')
const get = require('lodash/get')
const log = require('cozy-logger').namespace('e2e-onOperationOrBillCreate')

const assert = (cond, msg) => {
  if (!cond) {
    throw new Error(msg)
  }
}

const PREFIX = 'cozy-banks-e2e-onOperationOrBillCreate'

const dropData = async () => {
  log('info', 'Dropping data...')
  const doctypes = [
    'io.cozy.bank.operations',
    'io.cozy.bills',
    'io.cozy.bank.settings',
    'io.cozy.bank.accounts'
  ]
  return couch.dropDatabases(doctypes)
}

const isJSONFile = filename => filename.endsWith('.json')

const loadData = async () => {
  log('info', 'Loading data...')
  const dir = 'test/fixtures/matching-service'
  for (let fixture of fs.readdirSync(dir).filter(isJSONFile)) {
    ach(['import', path.join(dir, fixture)])
  }
}

const exportAndSnapshot = async () => {
  log('info', 'Exporting and snapshotting...')
  const exportFilename = `/tmp/${PREFIX}-export.json`
  ach(['export', 'io.cozy.bank.operations,io.cozy.bills', exportFilename])
  const actual = fs.readFileSync(exportFilename).toString()
  const testTitle = 'onOperationOrBillCreate'
  const filename = path.basename(__filename)
  const snapResult = toMatchSnapshot(actual, filename, testTitle)
  if (snapResult.pass) {
    log('info', 'Snapshot OK !')
  } else {
    console.error(snapResult.report()) // eslint-disable-line no-console
    throw new Error('Snapshot not OK')
  }
}

const checkSettings = async () => {
  const res = ach(['export', 'io.cozy.bank.settings'], {
    stdio: 'pipe'
  })
  const data = JSON.parse(res.stdout.toString())
  const settings = data['io.cozy.bank.settings']
  const settingDoc = settings[0]
  assert(settings.length == 1, 'There should be only 1 setting document')

  const definedPaths = [
    'appSuggestions.lastSeq',
    'billsMatching.transactionsLastSeq',
    'billsMatching.billsLastSeq',
    'notifications.lastSeq'
  ]
  for (let path of definedPaths) {
    assert(
      typeof get(settingDoc, path) !== 'undefined',
      `settings.${path} should not be undefined`
    )
  }
}

const testService = async options => {
  await dropData()
  await loadData()
  await runService('onOperationOrBillCreate', [JSON.stringify(options)])
  await exportAndSnapshot()
  await checkSettings()
}

const main = async () => {
  const e = await prompt('Clear data and run e2e test (y) ?')
  if (e !== 'y') {
    log('info', 'Aborting...')
    return
  }
  await makeToken(PREFIX)
  await testService({ transactionMatching: false })
  await testService({ billMatching: false })
}

main().catch(e => {
  console.error(e) // eslint-disable-line no-console
  process.exit(1)
})
