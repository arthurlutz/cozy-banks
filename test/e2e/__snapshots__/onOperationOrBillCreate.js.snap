// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`onOperationOrBillCreate 1`] = `
"{
  \\"io.cozy.bank.operations\\": [
    {
      \\"_id\\": \\"ikea_operation\\",
      \\"id\\": \\"ikea_operation\\",
      \\"account\\": \\"compteisa1\\",
      \\"amount\\": -15,
      \\"currency\\": \\"EUR\\",
      \\"date\\": \\"2018-10-17T00:00:00+02:00\\",
      \\"label\\": \\"IKEA\\",
      \\"manualCategoryId\\": \\"401060\\",
      \\"originalBankLabel\\": \\"IKEA\\",
      \\"type\\": \\"credit card\\",
      \\"bills\\": [
        \\"io.cozy.bills:ikea_bill\\"
      ]
    },
    {
      \\"_id\\": \\"remboursement_cpam\\",
      \\"id\\": \\"remboursement_cpam\\",
      \\"account\\": \\"compteisa1\\",
      \\"amount\\": 5,
      \\"currency\\": \\"EUR\\",
      \\"date\\": \\"2018-10-18T00:00:00+02:00\\",
      \\"label\\": \\"Remboursement CPAM\\",
      \\"manualCategoryId\\": \\"400610\\",
      \\"bills\\": [
        \\"io.cozy.bills:ameli_bill\\"
      ]
    },
    {
      \\"_id\\": \\"visite_medecin\\",
      \\"id\\": \\"visite_medecin\\",
      \\"account\\": \\"compteisa1\\",
      \\"amount\\": -20,
      \\"currency\\": \\"EUR\\",
      \\"date\\": \\"2018-10-18T00:00:00+02:00\\",
      \\"label\\": \\"Visite chez le m√©decin\\",
      \\"manualCategoryId\\": \\"400610\\",
      \\"bills\\": [
        \\"io.cozy.bills:ameli_bill\\"
      ],
      \\"reimbursements\\": [
        {
          \\"billId\\": \\"io.cozy.bills:ameli_bill\\",
          \\"amount\\": 5,
          \\"operationId\\": \\"remboursement_cpam\\"
        }
      ]
    }
  ],
  \\"io.cozy.bills\\": [
    {
      \\"_id\\": \\"ameli_bill\\",
      \\"amount\\": 5,
      \\"currency\\": \\"EUR\\",
      \\"date\\": \\"2018-10-18T00:00:00.000Z\\",
      \\"invoice\\": \\"io.cozy.files:ameli_invoice\\",
      \\"isRefund\\": true,
      \\"originalAmount\\": 20,
      \\"originalDate\\": \\"2018-10-18T00:00:00.000Z\\",
      \\"type\\": \\"health_costs\\",
      \\"vendor\\": \\"Ameli\\"
    },
    {
      \\"_id\\": \\"ikea_bill\\",
      \\"amount\\": 15,
      \\"date\\": \\"2018-10-17T00:00:00Z\\",
      \\"invoice\\": \\"io.cozy.files:ikea_invoice_id\\",
      \\"vendor\\": \\"Ikea\\"
    }
  ]
}"
`;
